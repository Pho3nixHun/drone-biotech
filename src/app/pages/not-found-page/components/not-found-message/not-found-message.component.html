<div class="text-center space-y-5 mx-5" *transloco="let t">
    <h1
        class="text-6xl sm:text-8xl font-bold text-center mb-4 w-full text-black"
    >
        {{ t(vm().titleKey) }}
    </h1>
    <p class="md:text-xl">
        {{ t(vm().descriptionKey) }}
    </p>
    <a [routerLink]="[vm().navItem.routerLink]" class="btn btn-secondary">{{
        t(vm().navItem.textKey)
    }}</a>
</div>
