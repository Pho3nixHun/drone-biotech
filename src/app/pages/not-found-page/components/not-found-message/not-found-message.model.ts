import { WithRouterLink } from '@interfaces/with-router-link.interface';
import { WithTextNode } from '@interfaces/with-text-node.interface';

interface NavItemVM extends WithRouterLink, WithTextNode {}

export interface NotFoundMessageVM {
    titleKey: string;
    descriptionKey: string;
    navItem: NavItemVM;
}
