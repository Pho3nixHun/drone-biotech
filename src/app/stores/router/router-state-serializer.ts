import {
    ActivatedRouteSnapshot,
    Data,
    Params,
    RouterStateSnapshot,
} from '@angular/router';
import { RouterStateSerializer } from '@ngrx/router-store';
import { RouterState } from './router.state';

export class CustomRouterStateSerializer
    implements RouterStateSerializer<RouterState>
{
    serialize = (state: RouterStateSnapshot): RouterState => ({
        url: state.url,
        params: mergeRouteParams(state.root, ({ params }) => params),
        queryParams: mergeRouteParams(
            state.root,
            ({ queryParams }) => queryParams
        ),
        data: mergeRouteData(state.root),
    });
}

const mergeRouteParams = (
    route: ActivatedRouteSnapshot,
    getter: (activatedRoute: ActivatedRouteSnapshot) => Params
): Params =>
    !route
        ? {}
        : {
              ...getter(route),
              ...mergeRouteParams(
                  (route.children.find(({ outlet }) => outlet === 'primary') ||
                      route.firstChild) as ActivatedRouteSnapshot,
                  getter
              ),
          };

const mergeRouteData = (route: ActivatedRouteSnapshot): Data =>
    !route
        ? {}
        : {
              ...route.data,
              ...mergeRouteData(
                  (route.children.find(({ outlet }) => outlet === 'primary') ||
                      route.firstChild) as ActivatedRouteSnapshot
              ),
          };
