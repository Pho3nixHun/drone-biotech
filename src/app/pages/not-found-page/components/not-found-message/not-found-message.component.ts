import { Component, input } from '@angular/core';
import { TranslocoModule } from '@jsverse/transloco';
import { NotFoundMessageVM } from './not-found-message.model';
import { RouterModule } from '@angular/router';

/**
 * NotFoundMessageComponent
 *
 * Type: Presentational
 *
 * Scope:
 * - Responsible for structuring a not found message component that contains dynamic content.
 * - Manages the overall layout and styling for the message component.
 *
 * Out-of-Scope:
 * - Not responsible for fetching or transforming data.
 *
 * Purpose (optional):
 * To provide a flexible and reusable not found message component layout.
 */
@Component({
    selector: 'app-not-found-message',
    imports: [TranslocoModule, RouterModule],
    templateUrl: './not-found-message.component.html'
})
export class NotFoundMessageComponent {
    public vm = input.required<NotFoundMessageVM>();
}
