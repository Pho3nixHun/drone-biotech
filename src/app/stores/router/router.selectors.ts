import { createFeatureSelector, createSelector } from '@ngrx/store';
import { RouterReducerState } from '@ngrx/router-store';
import { RouterState } from './router.state';

export const selectRouterState =
    createFeatureSelector<RouterReducerState<RouterState>>('router');

export const selectHeaderCanBeShown = createSelector(
    selectRouterState,
    (router) => router?.state.data['headerCanBeShown']
);

export const selectID = createSelector(
    selectRouterState,
    (router) => router?.state.params['id']
);
