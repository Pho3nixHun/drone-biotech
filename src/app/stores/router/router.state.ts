import { Data, Params } from '@angular/router';

export interface RouterState {
    url: string;
    queryParams: Params;
    params: Params;
    data: Data;
}
