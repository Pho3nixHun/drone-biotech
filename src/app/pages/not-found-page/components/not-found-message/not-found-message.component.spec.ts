import { ComponentFixture, TestBed } from '@angular/core/testing';
import { NotFoundMessageComponent } from './not-found-message.component';
import { Component } from '@angular/core';
import { NotFoundMessageVM } from './not-found-message.model';
import { getTranslocoModule } from 'transloco-testing.module';
import { provideRouter } from '@angular/router';

const enMock = {
    description: 'desc',
    text: 'txt',
    title: 'tit',
};

@Component({
    imports: [NotFoundMessageComponent],
    template: ` <app-not-found-message [vm]="vm"></app-not-found-message> `,
})
class TestHostComponent {
    public vm: NotFoundMessageVM = {
        descriptionKey: enMock.description,
        navItem: { textKey: enMock.text, routerLink: 'link' },
        titleKey: enMock.title,
    };
}

describe('NotFoundMessageComponent', () => {
    let fixture: ComponentFixture<TestHostComponent>;
    let compiled: HTMLElement;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [
                TestHostComponent,
                getTranslocoModule({
                    langs: { en: enMock },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
            providers: [provideRouter([])],
        }).compileComponents();

        fixture = TestBed.createComponent(TestHostComponent);
        compiled = fixture.debugElement.nativeElement;
    });

    // Snapshot test
    it('should render the template when the VM is provided', () => {
        //Arrange

        //Act
        fixture.detectChanges();

        //Assert
        expect(compiled).toMatchSnapshot();
    });
});
