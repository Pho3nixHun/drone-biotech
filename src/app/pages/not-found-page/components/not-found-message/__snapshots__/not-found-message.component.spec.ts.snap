// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`NotFoundMessageComponent should render the template when the VM is provided 1`] = `
<div
  id="root0"
  ng-version="20.1.7"
>
  <app-not-found-message>
    <div
      class="text-center space-y-5 mx-5"
    >
      <h1
        class="text-6xl sm:text-8xl font-bold text-center mb-4 w-full text-black"
      >
         tit 
      </h1>
      <p
        class="md:text-xl"
      >
         desc 
      </p>
      <a
        class="btn btn-secondary"
        href="/link"
      >
        txt
      </a>
    </div>
    <!--container-->
  </app-not-found-message>
</div>
`;
